horovod/common/operations.cc

```diff
@@ -148,19 +145,24 @@ OperationManager* CreateOperationManager(HorovodGlobalState& state) {
   std::vector<std::shared_ptr<AllgatherOp>> allgather_ops;
   std::vector<std::shared_ptr<BroadcastOp>> broadcast_ops;
 
+  if (state.controller->GetControllerType() == Controller::ControllerType::MPI){
```

Comment:
I think the better thing to do here is to decouple the `MPIContext` from the `MPIController`.  Even though there's no way to do it now, it might make sense in the future to use a non-MPI control plane with MPI data transfer.

What we can do is keep `mpi_context` as a global variable at the top of this file, but in the future we will only initialize it if using MPI for the control or data plane.  Then we can pass `mpi_context` to the `MPIController` constructor.

I would also pull the stuff in `MPIController::Initialize` and `MPIController::Finalize` into the `MPIContext` class.

Reply:
Yes. That would be a more elegant way to abstract controller. 

Already refracted.